### PR TITLE
net: context: move send_cb from net_context to net_pkt

### PR DESCRIPTION
**Pull request by mniestroj**
Each net_context_send*() call accepts send_cb and user_data as
parameters. Those parameters were saved in net_context structure during
each send request. There were two problems related to such behavior:

 * 'user_data' is shared for connect_cb and recv_cb, which means that
   it got overwritten during each net_context_send*() call. This means
   that net_context API consumer needed (in most cases) to always pass
   user_data that was previously used for net_context_recv(), even when
   send_cb was NULL.
 * 'send_cb' together with 'user_data' got overwritten during each
   net_context_send*() call, so there was no way of keeping track which
   packet was actually sent and which was not.

Move send_cb and user_data to net_pkt structure, so no undesirable
overwriting happens. This allows to keep track of each data portion that
is requested to be sent with net_context_send*() APIs. Additionally
user_data is no longer shared with the one passed to net_context_recv(),
which means that the same pointer doesn't have to be passed during each
subsequent net_context_send() call.

Simplify and cleanup code assuming previous behavior.

### CODE REVIEW

**Review by tbursztyka — CHANGES_REQUESTED** [review]
nack on that one.

This goes against the network stack architecture.

I don't see the problem with passing user_data on each send calls. It was meant to be like that.

**Comment by jukkar** [comment]
It is not clear what use case this PR would help to solve, could you elaborate this a bit more.

> 'user_data' is shared for connect_cb and recv_cb, which means that it got overwritten during each net_context_send*() call. 

You probably meant send_cb here instead of connect_cb. If this is a big problem, we could add second user_data pointer to net_context struct if needed. We have been trying to avoid adding more stuff to net_pkt because it uses more RAM.

**Comment by pfalcon** [comment]
@mniestroj: Echoing concerns from previous comments: net_context was effectively deemed to be internal implementation layer of network stack, sockets were deemed to be the public API. So, why do you need to use it in the first place, and how would you relate to the fact that if some other party also using it, you would be breaking them now?

Beyond that, there's a bunch of "unbeautifulness" in the underlying netstack implementation, but why patch it randomly? E.g. you patch net_context_send, but there's also net_context_recv with about the same issues. Is it worth it? Is risk worth it? I don't see anything with starting with `test:` included in the commits...

**Comment by mniestroj** [comment]
>This goes against the network stack architecture.
>
>I don't see the problem with passing user_data on each send calls. It was meant to be like that.

@tbursztyka Then what is the purpose of passing user_data and send_cb at all? Unfortunately it is not clear for me. After going through networking code for the last few days the only use case for send_cb + user_data would be to keeping track whether sending network packet was successful and at which point in time has that happened. See also below example with `net_context_send()` consumer code.

>It is not clear what use case this PR would help to solve, could you elaborate this a bit more.

@jukkar Sure. I was trying to improve ESP WiFi (NET_OFFLOAD) driver, because there are currently race conditions that need to be solved to further improve its reliability. Looking at the `.send()` and `.sendto()` implementations I have noticed that it might be good idea to simplify "packet send" code path and always use workqueue for actually sending (offloading) packet to ESP WiFi chip, regardless `timeout` passed as function parameter. Next issue was `socket->tx_pkt`, which should be somehow protected from two threads trying to access it simultaneously. Those things can be easily fixed by using `net_pkt->work`, as implemented  [here](https://github.com/mniestroj/zephyr/commit/3ad69242dae8fee1765b925d548f076b98bd50e4).

The problem that I encountered was that each `.send()` and `.sendto()` calls could have different `send_cb` and `user_data` parameters values. If that is true, then we need to add each `send_cb`+`user_data` pair to `net_pkt`, so we call the right ones when packet is actually being sent. Looking at the net_context and net_offload APIs, there was nothing being said that they are always the same during each call. I haven't even found a purpose of those. There is no clear way how to treat them, when should callback be executed, etc. Some drivers ignore `send_cb` (they do not call it at all), TCP stack executes callback whenever new packet is scheduled (not actually sent) and net_if layer calls net_context->cb with net_context->user_data whenever a packet is sent over interface. In summary: pretty inconsistent behaviour and no clear way what is the purpose of `send_cb`.

Most code that actually makes some use of `send_cb` suggest that this callback should be executed whenever a network packet is being sent over network interface. So changes in this PR were supposed to improve handling of `send_cb`, making sure that the same `send_cb`+`user_data` is used that was passed to `net_context_send*()` APIs. I might misunderstood intention of `send_cb`, in which case this PR might be invalid.

>You probably meant send_cb here instead of connect_cb. If this is a big problem, we could add second user_data pointer to net_context struct if needed. We have been trying to avoid adding more stuff to net_pkt because it uses more RAM.

@jukkar Well, user_data is shared across all 3 callbacks. I should probably rephrase that somehow.

>Echoing concerns from previous comments: net_context was effectively deemed to be internal implementation layer of network stack, sockets were deemed to be the public API. So, why do you need to use it in the first place, and how would you relate to the fact that if some other party also using it, you would be breaking them now?

@pfalcon I try to understand (and possibly improve) net_context from network drivers point of view. As I have written above, I started from WiFi ESP driver and ended up at strange (to me) behavior of `send_cb` handling at net_context layer.

I understand your concern about breaking API compatibility. Question is: is the current behavior of `send_cb` completely documented/defined, and if so, is this behavior properly enforced in current networking stack? Let's assume below user code:
```
net_context_send(ctx, buf, len, cb_1, K_NO_TIME, user_data_1);
net_context_send(ctx, buf, len, cb_2, K_NO_TIME, user_data_2);
```
According to `net_context_send()` documentation "After the network buffer is sent, a caller-supplied callback is called.". It means to me that first `cb_1(user_data_1)` is called when first packet is sent, then `cb_2(user_data_2)` is called when second packet is sent. This doesn't work like that currently, and this PR was supposed to fix that.

>Beyond that, there's a bunch of "unbeautifulness" in the underlying netstack implementation, but why patch it randomly? E.g. you patch net_context_send, but there's also net_context_recv with about the same issues. Is it worth it? Is risk worth it? I don't see anything with starting with test: included in the commits...

Basically I need to know how to handle `net_context_send` in NET_OFFLOAD drivers, such as ESP WiFi. Second reason is that why not fixing stuff in small steps, hence its easier to review? 

Thanks for noticing lack of tests. I will include some tests as far as we agree that the direction of those changes is good enough. Right now we are far from that yet.

**Comment by tbursztyka** [comment]
Trying to solve a hypothetical high level API issue from the drivers was, I think, your mistake.

In the first place, you should not be moving **anything** that belongs to net_context to net_pkt. That's 2 different architectural blocks. 

- net_pkt is about data buffer and metadata about l2 and l3
- while net_context is all about enabling user API

So a net_pkt might belong to a net_context, but no net_context stuff belongs to net_pkt, ever.

Nowadays, net_context is there to enable socket. There is no other API that will use net_context.
In early stage of this network stack, net_context used to be the API exposed to users, but now it's socket API.
(There are one sample as well as a few tests that still uses net_context, either because it was not ported to socket or just by convenience.) 

And if you look at socket, you'll see it's being used only once: on listen call.
All other calls make sure to pass ctx->user_data back. (It's actually commented to be that way, in socket code).

So there is no way send/sendto etc.. will have different user_data.

**Comment by mniestroj** [comment]
>Trying to solve a hypothetical high level API issue from the drivers was, I think, your mistake.

It is not hypothetical problem, if that behaviour is what directly comes from net_context API documentation. And this is not my problem only, because there are many existing drivers that handle net_context callbacks inconsistently.

> In the first place, you should not be moving **anything** that belongs to net_context to net_pkt. That's 2 different architectural blocks.
> 
> * net_pkt is about data buffer and metadata about l2 and l3
> * while net_context is all about enabling user API
> 
> So a net_pkt might belong to a net_context, but no net_context stuff belongs to net_pkt, ever.

Understood. But... There is already a `net_pkt->context` that points to `net_context`. If `net_context_send_cb_t send_cb;` is the problem, then lets make it context agnostic `net_pkt_send_cb_t send_cb;` and adjust `net_context` to make use of it.

>Nowadays, net_context is there to enable socket. There is no other API that will use net_context.
In early stage of this network stack, net_context used to be the API exposed to users, but now it's socket API.
(There are one sample as well as a few tests that still uses net_context, either because it was not ported to socket or just by convenience.)

That is not true. There are `shell_telnet`, `net_shell`, `dns`, `socks` and `log_backend_net` all using net_context API. Together with `sockets` this makes **6** in-tree modules.

>So there is no way send/sendto etc.. will have different user_data.

First of all there are *5* other in-tree modules using net_context API + some possible out-of-tree modules as well.

Second thing is that this should be directly written in net_context API. This is what (for example) NET_OFFLOAD driver writers look into when they receive `net_context_send_cb send_cb` as send/sendto function parameter and they want to figure out how to deal with it.

Last thing... if net_context API is supposed to be just a helper layer for sockets, then there should be a note/warning in net_context API to stop using it directly. And in the long-term net_context.h should be deprecated and hidden from application code at all.


With all this discussion, I still don't know what is the purpose of `net_context_send_cb_t send_cb` and how should I deal with it in network driver code to respect current "networking stack" / "networking architecture".

**Comment by jukkar** [comment]
> There are `shell_telnet`, `net_shell`, `dns`, `socks` and `log_backend_net` all using net_context API. Together with `sockets` this makes **6** in-tree modules.

Yes, that is correct. But as Tomasz said, many of these uses are because the code was not ported to use socket or just by convenience (net_context is async API which can be quite useful). Anyway, the point is that net_context API should only be used inside the net stack, not by application.

> if net_context API is supposed to be just a helper layer for sockets, then there should be a note/warning in net_context API to stop using it directly. And in the long-term net_context.h should be deprecated and hidden from application code at all.

This is already documented in the documentation https://docs.zephyrproject.org/latest/reference/networking/net_context.html
I agree that we could move the header files away from `include/net` so that those would not be so easily found by app devs. I don't think we cannot deprecate the net_context API as we use it internally.

> The problem that I encountered was that each `.send()` and `.sendto()` calls could have different `send_cb` and `user_data` parameters values.

I think that for the code using send_cb, the actual cb is most probably always the same. It is the user_data that might differ.

> pretty inconsistent behaviour and no clear way what is the purpose of `send_cb`.

I have to agree with you that the usage of `send_cb` is a bit thin, and it is not really utilized much. I just wonder should we stop using the callback then if there is no need for that, and document it properly. Question to all, would that make sense here? Is there proper use cases for the send_cb, and if there is none, then we could just remove that.
